datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreTypes.kt

```diff
@@ -125,15 +125,29 @@ internal fun Coding.toCodeType(): CodeType {
 }
 
 fun Type.valueOrCalculateValue(): Type {
-  return if (this.hasExtension()) {
-    this.extension
-      .firstOrNull { it.url == EXTENSION_CQF_CALCULATED_VALUE_URL }
-      ?.let { extension ->
-        val expression = (extension.value as Expression).expression
-        fhirPathEngine.evaluate(this, expression).singleOrNull()?.let { it as Type }
+  return if (getValueString(this) != null) {
+    this
+  } else {
+    this.cqfCalculatedValueExpression?.let { expression ->
+      try {
+        fhirPathEngine.evaluate(this, expression.expression).singleOrNull() as? Type
+      } catch (e: FHIRException) {
+        Timber.w("Could not evaluate expression with FHIRPathEngine", e)
+        null
       }
+    }
       ?: this
-  } else {
-    this
   }
 }
+
+internal val Type.isCqfCalculatedValue
+  get() = this.hasExtension(EXTENSION_CQF_CALCULATED_VALUE_URL)
+
+internal val Type.cqfCalculatedValueExpression
+  get() =
+    this.takeIf { isCqfCalculatedValue }
+      ?.getExtensionByUrl(EXTENSION_CQF_CALCULATED_VALUE_URL)
+      ?.value as? Expression
```

Comment:
can this be simplified? and the previous extension function can also be removed.

```suggestion
internal val Type.cqfCalculatedValueExpression
  get() =
    this
      ?.getExtensionByUrl(EXTENSION_CQF_CALCULATED_VALUE_URL)
      ?.value as? Expression
```